### PR TITLE
EM for GLM-HMM

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
## Summary

This PR refactors and significantly improves the GLM-HMM Expectation-Maximization (EM) algorithm implementation, with major improvements to code organization, performance, and test coverage.

  ## Key Changes

  ### GLM-HMM EM Algorithm Improvements
  - **Modularized EM implementation**: Refactored the EM algorithm for better maintainability and composability
  - **Improved parameter structure**: Changed from single `projection_weights` to tuple-based `glm_params` (coef, intercept) for clearer separation of concerns
  - **Enhanced convergence checking**: Added proper convergence detection with configurable tolerance and tracking of previous log-likelihoods
  - **Better JIT compilation**: Improved static argument handling and compilation behavior for faster execution
  - **Accurate likelihood computation**: Now uses actual data log-likelihood instead of approximate metrics

  ### Test Suite Reorganization
  - **Consolidated GLM-HMM tests**: Moved all GLM-HMM-specific tests to `test_glm_hmm_algorithms.py` (+2,421 lines)
  - **Extracted common regressor tests**: Created `test_base_regressor_subclasses.py` (+1,143 lines) to avoid duplication across GLM and GLM-HMM tests
  - **Reduced test file sizes**: Significantly streamlined `test_glm.py` (-1,296 lines net) by moving shared tests
  - **Removed obsolete tests**: Deleted old `test_glm_hmm.py` (-1,234 lines) in favor of the new comprehensive test suite
  - **Improved test performance**: Optimized test execution speed

  ### Code Architecture Improvements
  - **Generic base regressor**: Made `BaseRegressor` generic with `ParamsT` type parameter for better type safety
  - **Refactored parameter initialization**: Split `initialize_params()` into public API and private `_initialize_parameters()` for clearer responsibilities
  - **Reorganized utilities**: Moved `inverse_link_function_utils.py` from `glm/` to top-level `src/nemos/` for better accessibility
  - **Enhanced type casting**: Added utilities in `type_casting.py` for consistent JAX array conversion

  ### New Features
  - **GLM-HMM simulation script**: Added `scripts/generate_simulation_glm_hmm.py` for generating synthetic GLM-HMM datasets with configurable neurons and states
  - **Improved likelihood handling**: Added `prepare_likelihood_func()` for consistent likelihood function preparation across population and single-neuron GLMs

  ### Other Improvements
  - Updated regularizer handling and validation logic
  - Enhanced solver instantiation to accept loss functions directly
  - Various bug fixes and code quality improvements
  - Documentation updates

  ## Testing
  - All existing tests pass with the refactored implementation
  - New comprehensive test suite provides better coverage of GLM-HMM functionality
  - Tests now run faster due to optimizations

  ## Statistics
  - **29 files changed**: 4,868 insertions(+), 2,715 deletions(-)
  - Net addition of ~2,150 lines despite significant refactoring

  ## Breaking Changes
  ⚠️ **Parameter structure change**: GLM-HMM now uses `glm_params: Tuple[Array, Array]` instead of `projection_weights: Array`. Code using the GLM-HMM API will need to be updated accordingly.

### CODE REVIEW

File: src/nemos/glm_hmm/expectation_maximization.py [hunk 1]
```diff
@@ -652,6 +650,270 @@ def run_m_step(
...
+    tol : float
+        Absolute tolerance threshold.
+
+    Returns
+    -------
+    : Array
+        Boolean indicating convergence.
+    """
+    delta = jnp.abs(state.data_log_likelihood - state.previous_data_log_likelihood)
+    return delta < tol
```
Comment (BalzaniEdoardo) [inline src/nemos/glm_hmm/expectation_maximization.py]: @ahwillia @sjvenditto, this is the termination criteria I am currently setting as default for the EM but I don't think this is the best we can do. Do you have suggestions? Maybe a relative increment, or something more sophisticated?

File: src/nemos/glm_hmm/expectation_maximization.py [hunk 2]
```diff
@@ -652,6 +650,270 @@ def run_m_step(
...
+    y: Array,
+    inverse_link_function: Callable,
+    likelihood_func: Callable,
+    solver_run: Callable,
+    is_new_session: Array,
+) -> GLMHMMState:
+    """Single EM iteration step."""
+    previous_state = carry
+
+    (posteriors, joint_posterior, _, new_log_like, alphas, _) = forward_backward(
```
Comment (sjvenditto) [inline src/nemos/glm_hmm/expectation_maximization.py]:
```suggestion
    (posteriors, joint_posterior, _, new_log_like, _, _) = forward_backward(
```
`alphas` is not being used

**Review by sjvenditto — APPROVED** [review]
Everything looks good and clean!
